src/Avalonia.Controls/ToolTip.cs

```diff
@@ -297,26 +349,76 @@ private static void IsOpenChanged(AvaloniaPropertyChangedEventArgs e)
 
                 toolTip.AdornedControl = control;
                 toolTip.Open(control);
-                toolTip?.UpdatePseudoClasses(newValue);
             }
             else if (control.GetValue(ToolTipProperty) is { } toolTip)
             {
                 toolTip.AdornedControl = null;
                 toolTip.Close();
-                toolTip?.UpdatePseudoClasses(newValue);
             }
         }
 
-        IPopupHost? IPopupHostProvider.PopupHost => _popup?.Host;
+        private static bool IsOpenCoerce(AvaloniaObject avaloniaObject, bool isOpen)
+        {
+            if (avaloniaObject is Control control)
+            {
+                if (isOpen)
+                {
+                    var args = new RoutedEventArgs(ToolTipOpeningEvent);
+                    control.RaiseEvent(args);
```

Comment:
I'll let @grokys comment on this, but I'm a bit worried of having such side effects in coercion methods, potentially leading to reentrancy or other hard to debug problems. Coercion methods are called in the middle of setting an `EffectiveValue`, and I don't think this code currently handle reentrancy. (For example, `PropertyChanged` isn't called until all the internal effective value bookkeeping is done to avoid issues.)

Reply:
Yeah, didn't think about reentrancy. It's technically possible to change IsOpen value inside.

Reply:
It's also possible for the events to be raised when the value isn't changing. Aside from any internal coercion calls, the user can do this at any time:

```
myControl.CoerceValue(ToolTip.IsOpenProperty);
```

`ToolTip.IsOpenChanged` is the right place for this code. It's a bit weird that `IsOpen` will be true when `ToolTipOpening` is raised, but the event name makes it very clear what's going on.